### ISSUE TITLE
iOS WebView in a Layout does not size properly

### ISSUE THREAD

**Issue by czuck** [dotnet/maui#21604]
### Description

On Android the WebView is large enough to display the content:
![image](https://github.com/dotnet/maui/assets/1641428/d30d7345-e048-406d-9752-f4d7721e8278)

on iOS it is not:
![image](https://github.com/dotnet/maui/assets/1641428/695a5754-eca7-4179-95dc-29ca3c4aa257)

### Steps to Reproduce

Create a file > new Maui app.  Change the MainPage.xaml to 
```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:cleanApp="clr-namespace:CleanApp"
             x:Class="CleanApp.MainPage">
        <VerticalStackLayout BackgroundColor="BlueViolet">
            <Label x:Name="TitleLabel"
                   FontSize="16"
                   TextColor="Black"
                   BackgroundColor="Orange"
                   VerticalOptions="Start" />
            <WebView x:Name="WebViewDescription"
                     HorizontalOptions="Fill"
                     BackgroundColor="Orange"
                     AutomationProperties.IsInAccessibleTree="False" />
        </VerticalStackLayout>
</ContentPage>
```
and the MainPage.xaml.cs to:
```
namespace CleanApp
{
    public partial class MainPage : ContentPage
    {
        public const string HtmlPage = @"<!DOCTYPE html><html><head><meta name='viewport' content='width=device-width, initial-scale=1' /><style>{0}</style></head><body>{1}</body></html>";
        public const string MessageCss = "div.wysiwygtext p img {max-width: 100%;height: auto;}";

        public MainPage()
        {
            InitializeComponent();
            TitleLabel.Text = "Welcome to my web view test";
            WebViewDescription.Source = new HtmlWebViewSource { Html = string.Format(HtmlPage, MessageCss, "<div class=\"wysiwygtext\"><p>First Webview -  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p></div>") };
        }
    }

}
```
run the app.  It was found in version 8.0.14

### Version with bug

8.0.10 SR3

### Is this a regression from previous behavior?

Yes, this used to work in Xamarin.Forms

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

**Comment by AndreKraemer** [comment]
I encountered this issue myself while migrating my initial Xamarin.Forms applications to .NET MAUI. While it can be frustrating, it's important to note that this behavior is not a bug but rather by design as the layout logic for StackLayouts got changed between Xamarin.Forms and .NET MAUI. Referencing the [documentation](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/webview?view=net-maui-8.0&pivots=devices-android), we find the following crucial information:

> Important
> 
> A [WebView](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.webview) must specify its [HeightRequest](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.visualelement.heightrequest#microsoft-maui-controls-visualelement-heightrequest) and [WidthRequest](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.visualelement.widthrequest#microsoft-maui-controls-visualelement-widthrequest) properties when contained in a [HorizontalStackLayout](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.horizontalstacklayout), [StackLayout](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.stacklayout), or [VerticalStackLayout](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.verticalstacklayout). If you fail to specify these properties, the [WebView](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.webview) will not render.
> 

In addition to setting a specific `HeightRequest`, another solution is to utilize a `Grid` instead of a `VerticalStackLayout`:
```xml
    <Grid BackgroundColor="BlueViolet" RowDefinitions="auto, *">
        <Label
            x:Name="TitleLabel"
            BackgroundColor="Orange"
            FontSize="16"
            TextColor="Black"
            VerticalOptions="Start" />
        <WebView
            x:Name="WebViewDescription"
            Grid.Row="1"
            AutomationProperties.IsInAccessibleTree="False"
            BackgroundColor="Orange"
            HorizontalOptions="Fill" />
    </Grid>
```
However, it's important to note that while this solution addresses the layout issue, the WebView will now take all the available space on the screen and may grow larger than the content, resulting in a layout that differs from your Android screenshot.

**Comment by czuck** [comment]
Yes, it is very frustrating that Stacked Layouts were changed so they are practically unusable and grids don't fill the gap.  And the documentation is not accurate, the WebView IS rendered in both iOS and Android without either HeightRequest or WidthRequest being defined.  The real issue is the WebView isn't being resized once its content is loaded on iOS.  I was able to get around this using a custom handler, here it is in case anyone else can use it:
```
public partial class ExtendedWebViewHandler : WebViewHandler
{
    protected override void ConnectHandler(WKWebView platformView)
    {
        base.ConnectHandler(platformView);
        WebView.Loaded += OnLoaded;
    }


    protected override void DisconnectHandler(WKWebView platformView)
    {
        WebView.Loaded -= OnLoaded;
        base.DisconnectHandler(platformView);
    }

    WebView WebView => VirtualView as WebView;


    private void OnLoaded(object sender, EventArgs e)
    {
        PlatformView.NavigationDelegate = new ExtendedWKNavigationDelegate(WebView);

        // For fixing white flash issue in webview on dark themes/backgrounds and disable webview's scrolling
        PlatformView.Opaque = false;
        PlatformView.BackgroundColor = UIColor.Clear;
        PlatformView.ScrollView.ScrollEnabled = false;
    }

    class ExtendedWKNavigationDelegate: WKNavigationDelegate
    {
        private WebView _webView;
        public ExtendedWKNavigationDelegate(WebView webview)
        {
            _webView = webview;
        }
        public override async void DidFinishNavigation(WKWebView platformView, WKNavigation navigation)
        {
            while (platformView.IsLoading)
            {
                await Task.Delay(100); // wait here till content is rendered
            }
            _webView.HeightRequest = platformView.ScrollView.ContentSize.Height;
        }
    }
```

**Comment by ninachen03** [comment]
Can repro this issue at iOS platform on the latest 17.10.0 preview2 (8.0.14 & 8.0.7).
![image](https://github.com/dotnet/maui/assets/128687574/b3ddab5f-3d13-4a7b-b313-92247d29b162)

**Comment by thedee** [comment]
I have the same issue.  The problem with our situation is that we can't set a Height because the data is dynamic and can change.   So setting a HeightReqeust to 50 or whatever will not work.

@czuck I tried your example but for some reason the DidFinishNavigation is never firing.  This used to fire in Xamarin but in Maui.

**Comment by imoulas** [comment]
A decent workaround is to get the size of the page via javascript. 

You can add to your bottom of your web page the following code:
`
<div id=""page-height""></div>
<script>
      const body = document.body;
      const html = document.documentElement;
      const height = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight);     
      document.getElementById('page-height').innerHTML = height;
</script>
`

and after the page loads you can call the following code:

yourpage.xaml
`
 <WebView x:Name="yourwebview">
      <WebView.Source>
          <HtmlWebViewSource Html="{Binding exampleHtml}" />
      </WebView.Source>
  </WebView>
`

yourpage.xaml.cs
`	
var result = await yourwebview.EvaluateJavaScriptAsync($"document.getElementById('page-height').innerText");
Console.WriteLine(result);
yourwebview.HeightRequest = Convert.ToDouble(result.ToString());
`

I hope that helps.

**Comment by Ahamed-Ali** [comment]
Hi @czuck ,

Based on the documentation [link](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/webview?view=net-maui-9.0&pivots=devices-ios) ,  We must specify width and height values when webview is wrapped by StackLayout, VerticalStackLayout and so on.

![Image](https://github.com/user-attachments/assets/eb0ec3bf-ef29-484b-ab70-38fe4fe48033)

**Comment by czuck** [comment]
> Hi [@czuck](https://github.com/czuck) ,
> 
> Based on the documentation [link](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/webview?view=net-maui-9.0&pivots=devices-ios) , We must specify width and height values when webview is wrapped by StackLayout, VerticalStackLayout and so on.
> 
> ![Image](https://github.com/user-attachments/assets/eb0ec3bf-ef29-484b-ab70-38fe4fe48033)

As stated on April 2 above, " ..the WebView IS rendered in both iOS and Android without either HeightRequest or WidthRequest being defined. The real issue is the WebView isn't being resized once its content is loaded on iOS."  It is resized on android so this is an inconsistency between the platforms.

**Comment by czuck** [comment]
I'm having the same issue in a grid if the row height is set to Auto.

**Comment by PureWeen** [comment]
Duplicate of #10777
